docs: Add decision record for SOME/IP Gateway API (#234)
eclipse-score/inc_someip_gateway: docs/architecture/dec_someip_api_abstraction.rst

..
   # *******************************************************************************
   # Copyright (c) 2026 Contributors to the Eclipse Foundation
   #
   # See the NOTICE file(s) distributed with this work for additional
   # information regarding copyright ownership.
   #
   # This program and the accompanying materials are made available under the
   # terms of the Apache License Version 2.0 which is available at
   # https://www.apache.org/licenses/LICENSE-2.0
   #
   # SPDX-License-Identifier: Apache-2.0
   # *******************************************************************************

DR-001-Arch: Use socom as SOME/IP stack abstraction
===================================================

**Date:** 2026-08-11

.. dec_rec:: Use socom as SOME/IP stack abstraction
   :id: dec_rec__arch__use_socom_someip_abstraction
   :status: accepted
   :version: 1
   :context: SOME/IP Gateway
   :decision: Option 2 (socom binding)

Context / Problem
-----------------

The SOME/IP Gateway needs an abstraction layer to decouple from the specific SOME/IP stack implementation.

> In addition the SOME/IP Gateway is split into two processes (``someipd`` and ``gateway``) for safety reasons.

Reviewer: ```suggestion
In addition the SOME/IP Gateway is split into two processes (``someipd`` and ``gatewayd``) for safety reasons.
```

During the architecture discussion, two main reuse-oriented options were considered:

* Option 1: Implement the SOME/IP stack as a ``mw::com`` binding.
* Option 2: Implement the SOME/IP stack with its own SOME/IP abstraction.

The decision needed to balance implementation complexity, extensibility, and
ability to support gateway-specific runtime behavior.

Options Considered
------------------

Option 1: SOME/IP stack as ``mw::com`` binding
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This option would integrate SOME/IP through the ``mw::com`` binding API and
reuse ``mw::com`` as the central integration layer.

Pros:

* Already safety qualified
* Reuses existing ``mw::com`` integration concepts.
* Keeps one primary communication abstraction in the architecture.

Cons:

* ``mw::com`` is designed to know all bindings up front at compile time
  * To avoid a direct dependency on the SOME/IP stack an interface / abstraction would still be needed at this point
* Would need an additional control channel next to lola for getting event subscription state and FindService
  * Or add SOME/IP specific extensions to lola

> * lola is optimized for 1:n communication, but the gateway needs 1:1 communication

Reviewer: ```suggestion
* LoLa is optimized for 1:n communication, but the gateway needs 1:1 communication
```

* Configuring ``mw::com`` could become complex
  * Assuming that a SOME/IP service is mapped to a ``mw::com`` service
  * ``mw::com`` service configuration format might need SOME/IP specific extensions

>   * The ``mw::com`` configuration needs to be present up front at both ``someipd`` and ``gateway`` processes

Reviewer: ```suggestion
* ``mw::com`` is designed to know all bindings up front at compile time
* To avoid a direct dependency on the SOME/IP stack an interface / abstraction would still be needed at this point
* Would need an additional control channel next to LoLa for getting event subscription state and FindService
* Or add SOME/IP specific extensions to lola
* LoLa is optimized for 1:n communication, but the gateway needs 1:1 communication
* Configuring ``mw::com`` could become complex
* Assuming that a SOME/IP service is mapped to a ``mw::com`` service
* ``mw::com`` service configuration format might need SOME/IP specific extensions
* The ``mw::com`` configuration needs to be present up front at both ``someipd`` and ``gateway`` processes
```

Option 2: SOME/IP stack with its own SOME/IP abstraction
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This option uses ``socom`` as the abstraction used by the SOME/IP stack.

> For implementation of the IPC binding building blocks of `mw::com` are reused.

Reviewer: ```suggestion
For implementation of the IPC binding building blocks of ``mw::com`` are reused.
```

These are namely:

* ``message_passing``: for message transport between processes
* ``shared_memory``: for shared memory management at each process

Pros:

* ``socom`` provides a simpler API surface for this integration scenario.
* ``socom`` already supports registering bindings, which is needed by the
  selected architecture.
* ``socom`` supports enables a focused SDK-style integration for SOME/IP plugins.
* Keeps API for IPC (``mw::com``) and SOME/IP stack (``socom``) separate, which keeps the scope of each API clear.

Cons:

* Introduces another abstraction alongside ``mw::com`` and requires clear interface boundaries in documentation and code ownership.

Evaluation
----------

Option 1 would require missing capabilities and expected API extensions in
``mw::com`` before it can support the intended SOME/IP gateway behavior cleanly.

> To be oblivious of the concrete SOME/IP stack, ``mw::com`` would need to provide a suitable abstraction layer.

Reviewer: ```suggestion
To be independent of of the concrete SOME/IP stack, ``mw::com`` would need to provide a suitable abstraction layer.
```

or

```suggestion
To be agnostic to the concrete SOME/IP stack, ``mw::com`` would need to provide a suitable abstraction layer.
```

Option 2 provides the necessary extensibility (binding registration) with a
simpler integration model and method support out of the box, reducing the amount
of framework adaptation needed for the gateway implementation.

Decision
--------

Option 2 was selected: the SOME/IP stack abstraction in this module is based on
``socom``.

Consequences
------------

> * Gateway and daemon integration points are built around ``socom`` contracts.

Reviewer: ```suggestion
* The integration between the ``gateway`` and ``someipd`` daemon is based on ``socom`` interfaces.
```

..
   # *******************************************************************************
   # Copyright (c) 2026 Contributors to the Eclipse Foundation
   #
   # See the NOTICE file(s) distributed with this work for additional
   # information regarding copyright ownership.
   #
   # This program and the accompanying materials are made available under the
   # terms of the Apache License Version 2.0 which is available at
   # https://www.apache.org/licenses/LICENSE-2.0
   #
   # SPDX-License-Identifier: Apache-2.0
   # *******************************************************************************

DR-001-Arch: Use socom as SOME/IP stack abstraction
===================================================

**Date:** 2026-08-11

.. dec_rec:: Use socom as SOME/IP stack abstraction
   :id: dec_rec__arch__use_socom_someip_abstraction
   :status: accepted
   :version: 1
   :context: SOME/IP Gateway
   :decision: Option 2 (socom binding)

Context / Problem
-----------------

The SOME/IP Gateway needs an abstraction layer to decouple from the specific SOME/IP stack implementation.

> In addition the SOME/IP Gateway is split into two processes (``someipd`` and ``gatewayd``) for safety reasons.

Reviewer 2: ```suggestion
In addition the SOME/IP Gateway is split into two processes (``someipd`` and ``gatewayd``) for safety reasons, which then need to be connected using some IPC mechanism.
```

> The semantics of the abstraction layer needs to be transported via IPC to each process.

Reviewer 2: ```suggestion
The semantics of the abstraction layer need to be reflected via the IPC layer.
```

During the architecture discussion, two main reuse-oriented options were considered:

* Option 1: Implement the SOME/IP stack as a ``mw::com`` binding.
* Option 2: Implement the SOME/IP stack with its own SOME/IP abstraction.

> 

Reviewer 2: I think there are three orthogonal decisions:

1. Gateway vs binding architecture (i.e. whether to have the ASIL SOME/IP code in the application context or as a separate gateway process): This was actually already discussed before in the feature request. There it was decided to go with the gateway approach. This then came up again in our discussions and we said we will stick to the gateway decision for now and could think about binding later.
2. Which abstraction to use in order to interface with the SOME/IP stack / someipd.
3. Which IPC technology/API to use between ASIL and QM SOME/IP code (i.e. the gateway/binding and the network stack implementation). The initial pitch used mw::com, then we discussed whether this was appropriate for the use case.

I think what we discuss here is the second and third point resulting in decision for SOCOM API and SOCOM IPC bridge respectively.

For the Abstraction (second decision) we considered the following alternatives:
- Abstraction close to the vsomeip API:
  - Pro:
    - API design already mostly determined by vsomeip API
  - Con:
    - API not designed for zero-copy
    - API closely coupled with the exact stack
    - No abstraction of the service handling, all stack interaction logic ("Glue code") propagates into the gatewayd
    - Separation of "bridge" and "API" would still have to be designed
- Service-based abstraction stack-independent (SOCOM):
  - Pro:
    - All of the cons of previous alternative
    - API design already part of existing pitch
    - Proven in use at Aumovio
  - Con:
    - Complexity of the API (measure: reduce complexity by removing certain unneeded features)

For the IPC implementation (third decision) we considered the following alternatives:

- Use mw::com with GenericSkeleton.
  - Pro:
    - Safety-certified
  - Con:
    - Uses a lot of resources to support 1:N communication, but we have 1:1
    - Needs config and config can't be just injected via API
    - Event-driven use cases not so well supported
    - We would still need the SOCOM abstraction to wrap around the logic how mw::com services are used to do the actual service abstraction.
- Use pure message passing
  - Pro:
    - One mechanism that can handle everything
  - Con:
    - Will create problems with safety certification later on. On QNX the caller is blocked if the server is not ready to be called. ⇒ Can't be used in the safe execution phase.
    - Overhead for transferring the payload data via message passing
- Use combination of message passing, and shared memory
  - Pro:
    - Can transfer data within shared memory ⇒ Zero-copy
    - Can use safe mechanisms for signalling in the safe execution phase (e.g. Pulses)
  - Con:
    - Will add some effort in qualification

Author: In the IPC binding we use both `message_passing` and `shared_memory`. I now wonder, if you knew that.

The decision needed to balance implementation complexity, extensibility, and
ability to support gateway-specific runtime behavior.

Options Considered
------------------

Option 1: SOME/IP stack as ``mw::com`` binding
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This option would integrate SOME/IP through the ``mw::com`` binding API and
reuse ``mw::com`` as the central integration layer.

Pros:

* Already safety qualified
* Reuses existing ``mw::com`` integration concepts.
* Keeps one primary communication abstraction in the architecture.

Cons:

* ``mw::com`` is designed to know all bindings up front at compile time

  * To avoid a direct dependency on the SOME/IP stack an interface / abstraction would still be needed at this point

* Would need an additional control channel next to LoLa for getting event subscription state and FindService

  * Or add SOME/IP specific extensions to LoLa

* LoLa is optimized for 1:n communication, but the gateway needs 1:1 communication
* Configuring ``mw::com`` could become complex

  * Assuming that a SOME/IP service is mapped to a ``mw::com`` service
  * ``mw::com`` service configuration format might need SOME/IP specific extensions

>   * The ``mw::com`` configuration needs to be present up front at both ``someipd`` and ``gateway`` processes

Reviewer: ```suggestion
  * The ``mw::com`` configuration needs to be present up front at both ``someipd`` and ``gatewayd`` processes
```

Pros:

* ``socom`` provides a simpler API surface for this integration scenario.
* ``socom`` already supports registering bindings, which is needed by the
  selected architecture.
* ``socom`` supports enables a focused SDK-style integration for SOME/IP plugins.
* Keeps API for IPC (``mw::com``) and SOME/IP stack (``socom``) separate, which keeps the scope of each API clear.

Cons:

* Introduces another abstraction alongside ``mw::com`` and requires clear interface boundaries in documentation and code ownership.

Evaluation
----------

Option 1 would require missing capabilities and expected API extensions in
``mw::com`` before it can support the intended SOME/IP gateway behavior cleanly.
To be agnostic to the concrete SOME/IP stack, ``mw::com`` would need to provide a suitable abstraction layer.

Option 2 provides the necessary extensibility (binding registration) with a
simpler integration model and method support out of the box, reducing the amount
of framework adaptation needed for the gateway implementation.

Decision
--------

Option 2 was selected: the SOME/IP stack abstraction in this module is based on
``socom``.

Consequences
------------

> * The integration between the ``gateway`` and ``someipd`` daemon is based on ``socom`` interfaces.

Reviewer: ```suggestion
* The integration between the ``gatewayd`` and ``someipd`` daemon is based on ``socom`` interfaces.
```
